### PR TITLE
Cover Block: Fix regressions

### PR DESCRIPTION
**Pull request by WunderBart**
## Description

This PR fixes the following Cover block regressions:
- **Restore background dim styles for old markup**
   Fixes situations where Cover blocks published pre [#35065]() lost their background dim.
- **Sets 100% default background dim for solid colors/gradients and 50% for images**:
   Fixes situations where an Image transformed into Cover gets a 100% opaque overlay by default.

Updated tests:
 - Added E2E test for preserving the background dim when transformed from the Image block
 - ~Updated failing Cover block fixtures~ (see https://github.com/WordPress/gutenberg/pull/36406#issuecomment-969331791)
 
## How has this been tested?

The following scenarios for the Cover block should be verified:
1. When setting background color via the block placeholder, its default opacity is set to 100%
2. When setting image/video via the block placeholder, the default overlay color opacity is set to 50%
3. When transforming from the Image/video block, the default overlay opacity should be 50%
4. Default overlay opacity for blocks published with v11.7.1 and older should persist when viewed via this branch. This can be tested with the following steps:
   1. `git checkout v11.7.1`,
   2. Add the Cover block with Image, do not change the overlay color - it should be black with 50% opacity by default,
   3. Publish the page/post,
   4. `git checkout trunk`,
   5. Hard refresh the published page,
   6. Notice that the overlay color is gone,
   7. `git checkout fix/cover-block-regression`,
   8. Hard refresh again,
   9. Notice the overlay is correct now (as described in step 2).

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/cover/transforms.js
Comment (fullofcaffeine) [inline packages/block-library/src/cover/transforms.js]: Oh, so this is needed for the logic [here](https://github.com/WordPress/gutenberg/pull/36406/files#diff-c9b6422f332979151b7228eb2af812e84859338e6b4d5ef38e5341ef6e8f5a39R397) to take place, right? I wonder why it was missing? Maybe just an oversight?

Comment (WunderBart) [reply packages/block-library/src/cover/transforms.js]: > Oh, so this is needed for the logic here to take place, right?

Yep!

> I wonder why it was missing? Maybe just an oversight?

I'd say so, yeah 🤷

**Comment by ramonjd** [comment]
This works as advertised for me:

✅ When setting background color via the block placeholder, its default opacity is set to 100%
✅ When setting image/video via the block placeholder, the default overlay color opacity is set to 50%
✅ When transforming from the Image/video block, the default overlay opacity should be 50%

Here's how it looks on 11.7.1
11 7 1-cover

And on trunk:
trunk-cover

And the same post running this branch:
fix-regression-cover

It looks like the `inserting-blocks.test.js.snap` snapshot file needs to be regenerated maybe?

**Comment by Mamaduka** [comment]
Hi, @WunderBart

I see block validation errors when testing this branch with the TT2 theme.

## Screenshot
CleanShot 2021-11-12 at 15 50 31

File: packages/block-library/src/cover/edit.js
Comment (Mamaduka) [inline packages/block-library/src/cover/edit.js]: Let's include `dimRatio` in dependencies.

Setting attributes can cause an "undo trap." It's not visible with the Cover block, but it might be worth indicating that these attribute updates shouldn't create the "undo" history. More details: https://github.com/WordPress/gutenberg/pull/34565

Currently, I had to click three times to undo the Cover block insertion.

https://user-images.githubusercontent.com/240569/141464027-6624b3cf-810c-4d47-aaf2-b7231b11f172.mp4

File: packages/block-library/src/cover/block.json
Comment (glendaviesnz) [inline packages/block-library/src/cover/block.json]: Thanks for looking at this @WunderBart 🎉

For me turning off the default here causes any blocks created prior to this patch to break with validation errors - which is why "backgroundType":"image" needed to be added to all the fixtures I think. With a default set the attribute is never serialized into the block content.

I think if we went with this approach we would need to add a new deprecation that serialized the backgroundType into the content for all existing cover blocks. An alternative might be to [check for url instead](https://github.com/WordPress/gutenberg/pull/36406#discussion_r748924933).

File: packages/block-library/src/cover/edit.js
Comment (glendaviesnz) [inline packages/block-library/src/cover/edit.js]: I tried an alternative approach here of checking to see if url was set instead and it seemed to work for image, video, plain color background, and for transforming from image. This has the advantage of not needing to add deprecation so I wonder it that is a better way to fix this? I may be missing some edge cases in my testing of it though.

File: packages/block-library/src/cover/transforms.js
Comment (glendaviesnz) [reply packages/block-library/src/cover/transforms.js]: > I wonder why it was missing? Maybe just an oversight? 

Because the `image` type was the default it didn't need to be set previously

**Comment by glendaviesnz** [comment]
> I see block validation errors when testing this branch with the TT2 theme.

I think this is due to [this change](https://github.com/WordPress/gutenberg/pull/36406#pullrequestreview-805531043)

File: packages/block-library/src/cover/edit.js
Comment (glendaviesnz) [reply packages/block-library/src/cover/edit.js]: done, do you think we should move the withDispatch to a useDispatch as  well?

**Comment by glendaviesnz** [comment]
@WunderBart I pushed a change that adds a check for `url` so doesn't require the deprecation caused by removing `backgroundType` default. It seems to work with all the use cases - but there are a couple of failing fixtures due to a `dimRatio` of `Nan` being returned somewhere, but have run out of time to get to the bottom if it - will sort first thing tomorrow - but feel free to revert my commit if you see other issues with it.

**Comment by glendaviesnz** [comment]
> @WunderBart I pushed a change that adds a check for `url` so doesn't require the deprecation caused by removing `backgroundType` default. It seems to work with all the use cases - I updated the fixtures locally but there are a couple of failing fixtures due to a `dimRatio` of `Nan` being returned somewhere - so I suspect there is an edge case I have overlooked, but have run out of time to get to the bottom if it, so have not pushed those fixture updates - will sort first thing tomorrow - but feel free to revert my commit if you see other issues with it.

File: packages/block-library/src/cover/edit.js
Comment (Mamaduka) [reply packages/block-library/src/cover/edit.js]: Thanks, @glendaviesnz. I think that would be a good change. We're trying to slowly move away from HOCs.

**Comment by glendaviesnz** [comment]
@WunderBart I noticed when trying to sort the fixtures that the removal of the default for `dimRatio` also meant the need for a deprecation to be added in order to prevent block validation issues for some existing content. I took another look at it and have pushed a change that simplifies things a bit by putting back the attribute defaults, and instead sets the initial dimRatio in the transformations. This seems to work across all the options for adding a background, including transforming an image or video block.

**Comment by ramonjd** [comment]
Just retesting this after @glendaviesnz's change in 65b52906fb446ec4cfa31c439b025c38bc27a07f

The cover block works as expected. Fixture tests are passing for me, and also pasting in deprecated Cover Block code migrates correctly.

Transforming from an Image > Cover Block adds a default dimRatio of `50`.

Creating a Cover Block with media, i.e., an image or video, sets the dimRatio to `50`

With a color background, it sets the dimRatio to `100`

**Comment by WunderBart** [comment]
> @WunderBart I noticed when trying to sort the fixtures that the removal of the default for `dimRatio` also meant the need for a deprecation to be added in order to prevent block validation issues for some existing content. I took another look at it and have pushed a change that simplifies things a bit by putting back the attribute defaults, and instead sets the initial dimRatio in the transformations. This seems to work across all the options for adding a background, including transforming an image or video block.

Oh, this looks much better now, thanks! It does seem to cover all the testing scenarios and we don't need to regenerate the fixtures 💪 

@ramonjd thanks for testing! It seems we need another approval to unlock this PR though 🙏

**Comment by WunderBart** [comment]
@glendaviesnz one small issue that I've noticed - the title color is not properly set initially when setting a video as a Cover background. Once the opacity slider is moved, it seems to be detecting and seting the correct value though:

https://user-images.githubusercontent.com/1451471/141986458-9fcdda63-40af-4d91-9b5f-5ce75e57c09c.mov

Should we leave this PR as is and address this issue in a follow-up PR? Let me know what you think!

**Comment by glendaviesnz** [comment]
> @glendaviesnz one small issue that I've noticed - the title color is not properly set initially when setting a video as a Cover background. Once the opacity slider is moved, it seems to be detecting and seting the correct value though:
> Should we leave this PR as is and address this issue in a follow-up PR? Let me know what you think!

@WunderBart lets make this a separate issue - https://github.com/WordPress/gutenberg/issues/36549 - it seems related to the third party `fast-average-color` library that is being used to get the average background color - it doesn't seem to return a result when a video background is chosen, so lets not hold up this fix waiting for that to be resolved.

I will get someone to give us a final review and sign off on this today.

**Review by ramonjd — APPROVED** [review]
Another test for good measure 👍 

This PR satisfies the original motivation, i.e., the same cover block in:

**11.7.1** 

Opacity is applied:

cover-version-11 7 1

**Current trunk**

Opacity is not applied:

cover-trunk

**This branch**

Opacity returns!
cover-this-branch

Tests pass when I run:

`npm run fixtures:generate`

🎉

**Comment by andrewserong** [comment]
This PR was cherry picked into the GB 11.9.1 point release in https://github.com/WordPress/gutenberg/commit/c51a1539fc001f69df69f6f40686de826596e679

**Comment by noisysocks** [comment]
Just noting for posterity that I had to cherry-pick f2188840e857842570ec6edfd38a0574d66102e3 and e7242a8e30cd05b4499c1d14cb1e1b075aa90850 into `wp/trunk` in order to cherry-pick `a67553fe28c66e4bf7639e44f41d1f19861cd4a5` without conflicts.

**Comment by noisysocks** [comment]
Just noting for posterity that I had to cherry-pick #36312 and #36321 into `wp/trunk` in order to cherry-pick a683c28 without conflicts.

**Pull request merged**
